caas/kubernetes/provider/storage/metadata.go

```diff
@@ -0,0 +1,275 @@
+	StorageClassName string
+}
+
+// PreferredStorage defines a matcher that can inject it's own logic into the
+// preferred storage choice for a given storage class and decide if it's a match
+// to it's rules.
+type PreferredStorage interface {
+	// Name returns the name used to uniquely identify this preferred storage
+	// matcher.
+	Name() string
+
+	// Matches returns true if this matcher matches the supplied storage class.
+	Matches(*storagev1.StorageClass) bool
+}
+
+// PreferredStorageAnnotation is an implementation of PreferredStorage that
+// matches based on the supplied storage class containing at least one of the
+// annotations defined here.
+type PreferredStorageAnnotation struct {
+	// NameVal defines the value return by this structs Name() method
+	NameVal string
+
+	// Annotations defines one more annotations that are used in matching. If
+	// any of the annotations defined in this map exist in the supplied storage
+	// class then the match is considered true.
+	Annotations map[string]string
+}
+
+// PreferredStorageDefault is an implementation of PreferredStorage that returns
+// true if the supplied Kubernetes storage class is considered the default
+// storage class for the cluster.
+type PreferredStorageDefault struct {
+	// NameVal defines the value return by this structs Name() method
+	NameVal string
+}
+
+// PreferredStorageProvisioner is an implementation of PreferredStorage that
+// returns true if the supplied Kubernetes storage classes provisioner matches
+// this provisioner.
+type PreferredStorageProvisioner struct {
+	// NameVal defines the value return by this structs Name() method
+	NameVal string
+
+	// Provisioner is the string to match on the storage classess provisioner
+	// member.
+	Provisioner string
+}
+
+// PreferredStorageList defined an ordered list of PreferredStorage matches to
+// test a given storage class against. The position of PreferredStorage matches
+// matches indicating they're preference.
+type PreferredStorageList []PreferredStorage
+
+const (
+	operatorStorageClassAnnotationKey = "juju.is/operator-storage"
+	workloadStorageClassAnnotationKey = "juju.is/workload-storage"
+)
+
+// PreferredStorageForCloud returns a PreferredStorageList for the supplied
+// cloud. If no cloud is found matching then a default list is provided.
+func PreferredStorageForCloud(cloud string) PreferredStorageList {
+	switch cloud {
+
+	// Microk8s
+	case caas.K8sCloudMicrok8s:
+		return PreferredStorageList{
+			&PreferredStorageAnnotation{
+				NameVal: "storage-annotations",
+				Annotations: map[string]string{
+					operatorStorageClassAnnotationKey: "true",
+					workloadStorageClassAnnotationKey: "true",
+				},
+			},
+			&PreferredStorageProvisioner{
+				NameVal:     "hostpath",
+				Provisioner: "microk8s.io/hostpath",
+			},
+		}
+
+	// Azure
+	case caas.K8sCloudAzure:
+		return PreferredStorageList{
+			// Nominated
+			&PreferredStorageAnnotation{
+				NameVal: "storage-annotations",
```

Comment:
why set the name here? 
shouldn't `PreferredStorageAnnotation` and the others have the name "built-in", so in the case here, all the caller has to do is pass in the required annotations to use